merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,872 @@
+    /// a correct size, which is able to fit the hash set, is the caller's
+    /// responsibility.
+    ///
+    /// Calling it in async context (or anywhere where the underlying data can
+    /// be moved in memory) is certainly going to cause undefined behavior.
+    pub unsafe fn from_bytes_init(
+        bytes: &mut [u8],
+        capacity_indices: I,
+        capacity_values: I,
+        sequence_threshold: usize,
+    ) -> &mut Self {
+        let hash_set_ref = &mut *(bytes.as_mut_ptr() as *mut Self);
+
+        hash_set_ref.capacity_indices = capacity_indices;
+        hash_set_ref.capacity_values = capacity_values;
+        hash_set_ref.sequence_threshold = sequence_threshold;
+
+        let offset = mem::size_of::<Self>();
+        hash_set_ref.indices =
+            NonNull::new(bytes.as_mut_ptr().add(offset) as *mut Option<I>).unwrap();
+
+        let offset = offset
+            + (mem::size_of::<Option<I>>()
+                * usize::try_from(hash_set_ref.capacity_indices).unwrap());
+        hash_set_ref.values =
+            NonNull::new(bytes.as_mut_ptr().add(offset) as *mut Option<HashSetCell>).unwrap();
+
+        hash_set_ref
+    }
+
+    // Insert a value into the hash set
+    pub fn insert(&mut self, value: &BigUint, sequence_number: usize) -> Result<(), HashSetError> {
+        for i in
+            0..usize::try_from(self.capacity_values).map_err(|_| HashSetError::IntegerOverflow)?
+        {
+            let i = I::try_from(i).map_err(|_| HashSetError::IntegerOverflow)?;
+            let probe_index = (value.clone() + i.to_biguint().unwrap() * i.to_biguint().unwrap())
+                % self.capacity_values.to_biguint().unwrap();
+            let probe_index = probe_index.to_usize().unwrap();
+            let index_bucket = unsafe { &mut *self.indices.as_ptr().add(probe_index) };
+
+            match index_bucket {
+                // The visited hash set cell points to a value in the array.
+                Some(value_index) => {
+                    let value_bucket = unsafe {
+                        &mut *self.values.as_ptr().add(
+                            usize::try_from(*value_index).map_err(|_| HashSetError::UsizeConv)?,
+                        )
+                    };
+                    match value_bucket {
+                        // The cell in the value array is already taken.
+                        Some(value_bucket) => {
+                            // We can overwrite that cell only if the element
+                            // is expired - when the difference between its
+                            // sequence number and provided sequence number is
+                            // greater than the threshold.
+                            if let Some(other_sequence_number) = value_bucket.sequence_number {
+                                let sequence_diff =
+                                    sequence_number.checked_sub(other_sequence_number);
```

Comment:
Wouldn't it be more efficient to do just add the sequence threshold to the value_bucket.sequence_number during insertion. That way you would only have one arithmetic operation at insertion and just do comparisons here.

Reply:
I don't get how would we have "only one arithmetic operation". I think what you're proposing is moving it from here to the program code (it still needs to be done somewhere). And it feels like a bad API to me.

Reply:
I mean nullifier insertion into the Merkle tree.
In detail, when inserting the nullifier into the Merkle tree we will set value_bucket.sequence_number to the current sequence number by invoking mark_with_sequence_number.
My idea is to amend mark_with_sequence_number from:
```
        match element {
            Some(element) => {
                element.sequence_number = Some(sequence_number);
                Ok(())
            }
            None => Err(HashSetError::ElementDoesNotExist),
        }
```

to
```
        match element {
            Some(element) => {
                element.sequence_number = Some(sequence_number +  self.threshold);
                Ok(())
            }
            None => Err(HashSetError::ElementDoesNotExist),
        }
 ```
 
 
 Now you have only one arithmetic ops in mark_with_sequence_number.

Reply:
Ah, gotcha. Yes, that makes sense.